lib/elixir/src/elixir_module.erl

```diff
@@ -155,6 +159,16 @@ validate_inlines([Inline | Inlines], Defs, Unreachable, Acc) ->
   end;
 validate_inlines([], _Defs, _Unreachable, Acc) -> {ok, Acc}.
 
+validate_on_load_attribute({on_load, Def}, Defs, Unreachable, File, Line) ->
+  case lists:keyfind(Def, 1, Defs) of
+    false ->
+      elixir_errors:form_error([{line, Line}], File, ?MODULE, {undefined_on_load, Def});
+    _ ->
+      lists:member(Def, Unreachable) andalso
```

Comment:
Because you are looking up on unreachable, it means this code will work:

```elixir
defmodule Foo do
  @on_load :baz
  def bar, do: baz
  defp baz, do: :ok
end
```

So I think we should rather check the result returned by `lists:keyfind/3`. One of the elements in the tuple should say if it is `def`, `defmacro`, `defp` or `defmacrop`.